store/postgres/src/deployment.rs

```diff
@@ -251,6 +251,86 @@ pub fn features(conn: &PgConnection, site: &Site) -> Result<BTreeSet<SubgraphFea
         .collect()
 }
 
+pub fn transact_block(
+    conn: &PgConnection,
+    site: &Site,
+    ptr: &BlockPtr,
+    firehose_cursor: Option<&str>,
+    full_count_query: &str,
+    count: i32,
+) -> Result<(), StoreError> {
+    use crate::diesel::BoolExpressionMethods;
+    use subgraph_deployment as d;
+
+    // Work around a Diesel issue with serializing BigDecimals to numeric
+    let number = format!("{}::numeric", ptr.number);
+
+    // The big complication in this query is how to determine what the
+    // new entityCount should be. We want to make sure that if the entityCount
+    // is NULL or the special value `-1`, it gets recomputed. Using `-1` here
+    // makes it possible to manually set the `entityCount` to that value
+    // to force a recount; setting it to `NULL` is not desirable since
+    // `entityCount` on the GraphQL level is not nullable, and so setting
+    // `entityCount` to `NULL` could cause errors at that layer; temporarily
+    // returning `-1` is more palatable. To be exact, recounts have to be
+    // done here, from the subgraph writer.
+    //
+    // The first argument of `coalesce` will be `NULL` if the entity count
+    // is `NULL` or `-1`, forcing `coalesce` to evaluate its second
+    // argument, the query to count entities. In all other cases,
+    // `coalesce` does not evaluate its second argument
+    let count_update = if count == 0 {
+        // This amounts to a noop - the entity count does not change
+        "entity_count".to_string()
+    } else {
+        format!(
+            "coalesce((nullif(entity_count, -1)) + ({count}),
+                  ({full_count_query}))",
+            full_count_query = full_count_query,
+            count = count
+        )
+    };
```

Comment:
This duplication with `fn update_entity_count` is unfortunate, would it be possible to not duplicate this code?

Reply:
I factored that scary comment and string generation into a separate function.